test/irmin/test_tree.ml

```diff
@@ -571,92 +591,128 @@ let test_fold_force _ () =
 
   Lwt.return_unit
 
-let test_shallow _ () =
-  let* () =
-    let compute_hash ~subtree =
-      Tree.(add_tree empty) [ "key" ] subtree >|= Tree.hash
-    in
-    let leaf = Tree.of_concrete (c "0") in
-    let* shallow_leaf =
-      let+ repo = Store.Repo.v (Irmin_mem.config ()) in
-      Tree.shallow repo (`Contents (Tree.hash leaf, Metadata.default))
-    in
-    let* hash = compute_hash ~subtree:leaf in
-    let+ hash_shallow = compute_hash ~subtree:shallow_leaf in
-    Alcotest.(gcheck Store.Hash.t)
-      "Hashing a shallow contents value is equivalent to hashing the \
-       non-shallow contents"
-      hash hash_shallow
-  in
-  Lwt.return_unit
+(* Tests of "broken" trees: trees that can't be dereferenced. Tree currently
+      supports two varieties of broken tree:
 
-let test_dangling_hash _ () =
-  let check_exn pos f =
-    Lwt.catch
-      (fun () ->
-        let* _ = f () in
-        Alcotest.failf ~pos
-          "Expected a `Dangling_hash` exception, but no exception was raised.")
-      (function Tree.Dangling_hash _ -> Lwt.return_unit | exn -> Lwt.fail exn)
-  in
-  let* shallow_leaf, shallow_node =
+   - shallow trees containing [(repo, key)] pairs for which [repo] doesn't
+     contain [key]. Attempted dereferences should raise [Dangling_hash].
+
+   - pruned trees (hash-only tree nodes, with no underlying repository).
+     Attempting dereferences should raise [Pruned_hash]. *)
+module Broken = struct
+  let shallow_of_ptr kinded_hash =
     let+ repo = Store.Repo.v (Irmin_mem.config ()) in
```

Comment:
not related to this PR, but we don't close this anywhere, and maybe this causes side effects between tests? (if it's an issue we can also fix this in another PR)